### DISCUSSION TITLE
Show & Tell: Phase 19.4 — MultiModalEncoder: Unified Embedding Space & Cross-Modal Fusion

### DISCUSSION THREAD

**Discussion by web3guru888** [web3guru888/asi-build#475] [Show and tell]
## 🎛️ MultiModalEncoder — Unified Embedding Space & Cross-Modal Fusion

Phase 19.4 introduces **MultiModalEncoder**, the component that maps heterogeneous input modalities — text, images, audio, and future sensor data — into a **single shared vector space**. Downstream consumers receive a uniform `MultiModalEmbedding` regardless of source modality, enabling cross-modal similarity search, novelty detection, and grounded reasoning.

---

### Architecture Overview

MultiModalEncoder follows a **fan-in fusion** pattern:

1. **Per-modality encoders** — each `Modality` variant has a dedicated `EncoderBackend` that converts raw input into a fixed-dimensional vector.
2. **FusionLayer** — when multiple modalities arrive simultaneously, the fusion layer combines per-modality vectors into one embedding using the configured `FusionStrategy`.
3. **L2 normalisation** — the final embedding is L2-normalised so cosine similarity reduces to a simple dot product.

```
Text ──→ TextEncoder ──────┐
Image ─→ ImageEncoder ─────┼──→ FusionLayer ──→ MultiModalEmbedding
Audio ─→ AudioEncoder ─────┘       ↓
                           (ATTENTION / AVG / CONCAT)
                                   ↓
                           L2-normalise → unit vector
```

---

### Modality Enum & Backend Table

| Modality  | Default Backend    | Input Type        | Notes                                  |
|-----------|--------------------|-------------------|----------------------------------------|
| `TEXT`    | `SIMPLE_HASH`      | `str`             | Deterministic hash → bucket embedding  |
| `IMAGE`   | `CNN`              | `bytes` (pixels)  | Pixel grid → flatten → linear project  |
| `AUDIO`   | `SPECTROGRAM`      | `bytes` (PCM)     | FFT → mel-spectrogram → linear project |
| `VIDEO`   | *(future)*         | —                 | Planned: keyframe sampling + temporal  |
| `SENSOR`  | *(future)*         | —                 | Planned: time-series → 1-D conv        |

**EncoderBackend details:**

| Backend         | Strategy                                         | Trainable? | Use Case               |
|-----------------|--------------------------------------------------|------------|------------------------|
| `SIMPLE_HASH`   | Deterministic hash → distribute across dim buckets → normalise | No         | Baseline / testing      |
| `TRANSFORMER`   | Pre-trained transformer encoder (future)         | Yes        | Production text/NLU     |
| `CNN`           | Pixel flatten → learned linear projection        | Yes        | Image encoding          |
| `SPECTROGRAM`   | FFT → mel-spectrogram → learned linear projection | Yes       | Audio encoding          |

---

### FusionStrategy Comparison

| Strategy        | Formula                                    | Output Dim       | Pros                                  | Cons                                  |
|-----------------|--------------------------------------------|------------------|---------------------------------------|---------------------------------------|
| `CONCATENATE`   | `[v_text ‖ v_image ‖ v_audio]`            | `n × embed_dim`  | Preserves all info, no info loss      | Output dim grows with modality count  |
| `ATTENTION`     | `Σ αᵢ vᵢ` where `αᵢ = softmax(vᵢ · q)` | `embed_dim`      | Learned modality importance           | Requires query vector, more compute   |
| `AVERAGE`       | `(1/n) Σ vᵢ`                              | `embed_dim`      | Simple, fixed dim, no parameters      | Equal weight — ignores salience       |
| `WEIGHTED_SUM`  | `Σ wᵢ vᵢ` where `Σwᵢ = 1.0`             | `embed_dim`      | Configurable per-modality importance  | Manual weight tuning needed           |

---

### L2 Normalisation & Cosine Similarity

All embeddings are L2-normalised to unit length:

```
v̂ = v / ‖v‖₂
```

This means:
- **Cosine similarity** = dot product: `sim(a, b) = v̂ₐ · v̂ᵦ`
- **Range**: `[-1, 1]` — 1.0 = identical direction, 0.0 = orthogonal, -1.0 = opposite
- **Cross-modal search**: `encode("sunset photo") · encode(image_bytes)` yields meaningful similarity

---

### Integration Map

```
SemanticParser 19.1 ──(text intent)──→ MultiModalEncoder 19.4
                                              ↓
                                     MultiModalEmbedding
                                        ↙         ↘
              WorldModel 13.1 ←──(grounded state)   CuriosityModule 13.3
              (multi-modal world repr)               (embedding distance
                                                      → novelty score)
```

| Consumer              | How It Uses MultiModalEncoder                                      |
|-----------------------|---------------------------------------------------------------------|
| **SemanticParser 19.1** | Sends text to `encode_text()` → intent matching via cosine sim    |
| **WorldModel 13.1**    | Calls `encode([text, image])` → fused embedding for world state   |
| **CuriosityModule 13.3** | Computes embedding distance between new input and memory → novelty signal |
| **MemoryConsolidator 18.2** | Stores embeddings as `SemanticPattern` vectors for consolidation |

---

### Prometheus Metrics

| Metric                                      | Type      | Description                                 |
|---------------------------------------------|-----------|---------------------------------------------|
| `asi_multimodal_encode_total`               | Counter   | Total encode calls (label: modality)        |
| `asi_multimodal_encode_seconds`             | Histogram | Encode latency per call (label: modality)   |
| `asi_multimodal_fusion_total`               | Counter   | Fusion invocations (label: strategy)        |
| `asi_multimodal_embedding_dim`              | Gauge     | Configured embedding dimension              |
| `asi_multimodal_unsupported_modality_total` | Counter   | Rejected unsupported modality attempts      |

**PromQL examples:**

```promql
# Average encode latency by modality (5m window)
rate(asi_multimodal_encode_seconds_sum[5m]) / rate(asi_multimodal_encode_seconds_count[5m])

# Unsupported modality error rate
rate(asi_multimodal_unsupported_modality_total[5m])
```

**Grafana alert YAML:**

```yaml
- alert: MultiModalEncoderHighLatency
  expr: |
    rate(asi_multimodal_encode_seconds_sum{modality="IMAGE"}[5m])
    / rate(asi_multimodal_encode_seconds_count{modality="IMAGE"}[5m]) > 0.5
  for: 5m
  labels:
    severity: warning
  annotations:
    summary: "Image encoding latency > 500ms"
```

---

### 🔮 Open Questions

1. **Pre-trained weight loading** — Should MultiModalEncoder support loading pre-trained transformer/CNN weights at init? If so, what serialisation format (safetensors, ONNX, pickle)?
2. **Video modality** — How should video be decomposed? Keyframe sampling at N fps + temporal pooling, or full frame-by-frame encoding with a temporal transformer?
3. **Embedding drift detection** — Over time, distribution of embeddings may shift. Should we track centroid drift per modality and alert when cosine distance from baseline exceeds a threshold?

---

*Spec: [Phase 19.4 — MultiModalEncoder](#) · Issue: #481 · Wiki: Phase-19-MultiModal-Encoder*
